### PR TITLE
FilePersistence

### PR DESCRIPTION
**Pull request by Varadiell**
Edited files history is now saved in a SQLite3 database, in order to keep everything and have a better overview of the changes.

### CODE REVIEW

**Comment by Varadiell** [comment]
Quelques bugs sont toujours présents cependant... ceci n'est pas fonctionnel.

File: public/index.html [hunk 1]
```diff
@@ -34,26 +38,55 @@ <h4 class="col-xs-12">File History {{ history.visibility.states }}</h4>
...
+
         </div>
-      </div><!-- / -->
+
+      </div><!-- /history -->
+
     </div><!-- /container -->
 
+
+    <!-- DEBUT HISTORIQUE PERSISTENCE -->
```
Comment (FGRibreau) [inline public/index.html:66]: write everything in english :)

File: public/index.html [hunk 2]
```diff
@@ -34,26 +38,55 @@ <h4 class="col-xs-12">File History {{ history.visibility.states }}</h4>
...
+    <div ng-controller="HistoryPersistenceCtrl as historyPersistenceC">
```
Comment (FGRibreau) [inline public/index.html:67]: `historyPersistenceC` -> `historyPersistence`

File: public/index.html [hunk 3]
```diff
@@ -34,26 +38,55 @@ <h4 class="col-xs-12">File History {{ history.visibility.states }}</h4>
...
+      <div ng-repeat="historyP in historyPersistenceC.historyP" class="animate">
```
Comment (FGRibreau) [inline public/index.html:68]: `historyP in historyPersistenceC.historyP` -> `history in historyPersistence.history`

File: server.js [hunk 4]
```diff
@@ -7,7 +7,33 @@ var app = express();
...
-var config = require('./config')(logger);
+var config = require('./server/config')(logger);
+
+// History Persistence
+var sqlite3 = require('sqlite3').verbose();
+var db = new sqlite3.Database('dbHistoryPersistence.db');
+var historyPersistence = require('./server/HistoryPersistenceCtrl');
+
+// Database initialization
+db.serialize(function() {
```
Comment (FGRibreau) [inline server.js:18]: extract L.13-L.36 that into its own file

File: server.js [hunk 5]
```diff
@@ -29,15 +55,20 @@ sio.set('authorization', function (handshakeData, accept) {
...
       return socket.emit('error:auth', 'Unauthorized :)');
     }
-
-    // forward the event to everyone
-    sio.emit.apply(sio, ['file:changed'].concat(_.toArray(arguments)));
+    // Save the diff for history persistence
+    historyPersistence.persist(filePath, timestamp, content, function (err) {
+      if (err) {
+        console.error(err);
+        return socket.emit('file:changed:error', err);
```
Comment (FGRibreau) [inline server.js:67]: :+1:

File: server/HistoryPersistence.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+module.exports = function (sqlite) {
+
+  var historyP = [];
```
Comment (FGRibreau) [inline server/HistoryPersistence.js:5]: `historyEntries` maybe?

File: server/HistoryPersistence.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     * [persist description]
+     * @param  {String} filePath
+     * @param  {Datetime} timestamp
+     * @param  {String} content
+     * @param  {Function} f(err)
+     */
+    persist: function (filePath, timestamp, content, f) {
+      // SQLite request in order to add the diff to the database
+      var stmt = db.prepare("INSERT INTO historyDiff (filePath, timestamp, content) VALUES (?, ?, ?)");
```
Comment (FGRibreau) [inline server/HistoryPersistence.js:17]: historyDiff -> fileDiff

File: server/HistoryPersistence.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+     * @param  {Number} count  (MAX: 10)
+     * @param  {Function} f(err, Array[Object]);
+     * e.g.
+     *  [{
+     *    filePath: '',
+     *    timestamp: 120930293,
+     *    content: ''
+     *  }, ...]
+     */
+    getHistory: function (limit, offset, f) {
```
Comment (FGRibreau) [inline server/HistoryPersistence.js:33]: getFileDiffs ?

File: server/HistoryPersistence.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
+'use strict';
```
Comment (FGRibreau) [inline server/HistoryPersistence.js:1]: HistoryPersistence.js -> FileDiffPersistence.js

File: server/HistoryPersistenceCtrl.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+angular.module('FileSync').controller('HistoryPersistenceCtrl', ['HistoryPersistence',
+  function (HistoryPersistence) {
+  	var limit = 10;
```
Comment (FGRibreau) [inline server/HistoryPersistenceCtrl.js:5]: this is a constant:  `limit` -> `LIMIT`

File: server/HistoryPersistence.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+     *    timestamp: 120930293,
+     *    content: ''
+     *  }, ...]
+     */
+    getHistory: function (limit, offset, f) {
+      /* Requete SQLite */
+      limit = (limit > 10 || limit < 0) ? 10 : limit;
+      var stmt = db.prepare("SELECT * FROM historyDiff LIMIT ? OFFSET ?");
+      stmt.run(limit, offset);
+      historyP = stmt.finalize();
```
Comment (FGRibreau) [inline server/HistoryPersistence.js:38]: are you sure this is synchronous?

File: server/HistoryPersistenceCtrl.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+  	var offset = 0;
+
+    this.historyP = HistoryPersistence.getHistory(limit, offset, function (err) {
```
Comment (FGRibreau) [inline server/HistoryPersistenceCtrl.js:8]: this won't work

**Pull request closed without merging**